### ISSUE TITLE
Modal dialog and go back

### ISSUE THREAD

**Issue by collimarco** [turbolinks/turbolinks#561]
I have a simple use case: there is a [page](https://buonmenu.com/example) with a list of restaurant menus. When someone clicks on a menu that menu is opened, as a modal dialog, **without loading a new page**. Now I would like to close the menu / modal dialog when someone clicks Back in their browser.

With the history api is quite simple: you just need to `pushstate` when a menu is clicked. Then restore the page state listening to `popstate`.

The problem is that this seems impossible to do if I keep using Turbolinks, because it interferes with the history api.

**Is there any documentation that describes how to handle modal dialogs with Turbolinks?** 

Basically you can consider to have a page that can open some modal dialogs, lightboxes, etc. and when you click Back in the browser, it simply closes the dialog. How to achieve that if you have Turbolinks?

**Comment by welearnednothing** [comment]
Hey @collimarco,

You can listen to the `turbolinks:before-cache` event as explained in the "[Preparing the Page to be Cached](https://github.com/turbolinks/turbolinks#preparing-the-page-to-be-cached)" section of the README and use that as an opportunity to close the menu before leaving the page.

Taking a look at the code on the example page you provided, you should be able to do something like:

```
document.addEventListener("turbolinks:before-cache", function(evt) {
  openMenuNav(evt);
});
```

For anyone using [StimulusJS](https://stimulusjs.org), this can just be dropped into the `disconnect()` method of a controller. So if you have a MenuController, you could add something like:

```
// menu_controller.js
import { Controller } from 'stimulus'

export default class extends Controller {
  open() {
    ...
  }

  close() {
    ...
  }

  disconnect() {
    this.close()
  }
}
...
```

**Comment by collimarco** [comment]
@welearnednothing Thanks for the reply! However adding that code doesn't do anything... Probably you mean to add that together with something else, but what code?

Basically on the page I have this:

1. you see a list of restaurant menus: I call openMenu(event, menuId) when the user clicks one and that menu is displayed (changing the `display` property in css)
2. you are now inside a restaurant menu: I call openMenuNav(event) to go back and see all restaurant menus

Since in my site the term "menu" might be confusing, because we are talking about restaurants, this is the more general problem, that seems a common scenario:

1. you have a list of items on a page: you call openItem(event, id) when the user clicks an item and that item is displayed over the other content, like a modal (changing the `display` property in css)
2. you have now an item open (as a modal): when the user clicks the **close button** or **Go back** in the browser the modal must be closed
3. The user can continue to open and close other items and things should not break

Usually I would use the History API to add a page visit when an item is opened, so that I can then intercept the *Go back* event (using *popstate*). However Turbolinks already uses the History API in its own way and thus I should not interact with the History API directly. Any idea?

**Comment by domchristie** [comment]
Hi @collimarco 
This has been questioned a few times (#549, #491, #219, and #61), and there is a workaround (mentioned in those issues), however, it looks like this might break when the next version of Turbolinks is released.
Keeping #61 open to discuss this feature, but closing this one for now.

**Comment by welearnednothing** [comment]
@collimarco Ahhhh, I totally misread the problem. I read it through the lens of a problem I had solved that day. My bad, sorry for leading you astray! Thanks @domchristie for pointing in the right direction.
